### PR TITLE
Add files via upload

### PR DESCRIPTION
**Pull request by GlazerMann**
Lots of bug fixes

### CODE REVIEW

File: DOItools.php [hunk 1]
```diff
@@ -568,7 +578,7 @@ function checkTextForMetas($text){
     $authors = formatAuthors($author_list, true);
     foreach ($authors as $no => $auth) {
       $names = explode (', ', $auth);
-      $newp["author" . ($no + 1)] = $names[0];
```
Comment (ms609) [inline DOItools.php]: Are you sure it's safe to remove this?

**Comment by ms609** [comment]
Looks great, thanks... not quite clear that removing 'author' on line 571 is an unambiguously safe thing to do, but let me know if you think it is.

File: DOItools.php (hunk 1 above)
Comment (GlazerMann) [inline DOItools.php]: See line below where is it replaced with "last="

Comment (GlazerMann) [inline DOItools.php]: last and author are aliases, so it should be fine.  People prefer last when there is a first, and it is split in this case.

**Comment by GlazerMann** [comment]
Last and author are identical in template, do it is just to make human editors hsppy

**Comment by ms609** [comment]
Makes sense, thanks.  I've merged.

**Comment by GlazerMann** [comment]
We probably need to get this pushed up to the development version for actual use and testing.

Kurt



________________________________

**Comment by kaldari** [comment]
@GlazerMann: This patch breaks Citation bot. Specifically running https://tools.wmflabs.org/citations/doibot.php no longer works.

File: Template.php [hunk 2]
```diff
@@ -405,6 +406,26 @@ public function add_if_new($param, $value) {
...
+      return false;
+      case 'volume':
+        if ($this->blank($param)) {
+          if (  $this->get('journal') == "ZookKeys" ) add_if_new('issue',$value) ; // This journal has no volume
+          return $this->add($param, $value);
+        }
+      return false;
+      case 'bibcode':
+        if ($this->blank($param)) { 
+          $bibcode_pad =  strlen($value) - 19;
```
Comment (kaldari) [inline Template.php]: This should be 19 - strlen($value), not strlen($value) - 19.

File: Template.php [hunk 3]
```diff
@@ -405,6 +406,26 @@ public function add_if_new($param, $value) {
...
+          if($bibcode_pad > 0 ) {  // Paranoid, don't want a negative value, if bibcodes get longer
+              value = $value . str_repeat( ".", $bibcode_pad);  // Add back on trailing periods
```
Comment (kaldari) [inline Template.php]: Basic syntax error here. Should be $value.

**Pull request merged**